### PR TITLE
Improve MPI detection

### LINKED ISSUE #7045: Using Intel MPI with GCC
**Describe the bug**
It seems that in Meson 0.54.1, it is not possible anymore to use the mpi dependency to detect IntelMPI when using gcc.
This worked in 0.53 and is probably due to the change in behavior in baca2cd07d632dd5d944b2b364a297a0adf42028

**To Reproduce**
```
project('mpitest', 'c')
dependency('mpi')
```
with Intel `mpicc` in `$PATH`.

1. run meson with CC=icc: succeeds
2. run meson with CC=gcc
```
mpicc found: NO
Run-time dependency MPI found: NO (tried pkgconfig and config-tool)
```

**Expected behavior**
Meson should use the available `mpicc` to find the correct MPI implementation independent of the compiler.
I am also not sure if the comment
`# OpenMPI, which doesn't work with intel` in the code is accurate. I remember using OpenMPI with icc working just fine. Detecting whether we have Intel or OpenMPI based on compiler type therefore does not work.

**system parameters**
* native build
* CentOS 7.7
* Python 3.7.3
* Meson 0.45.1
* 1.9.0.git.kitware.dyndep-1.jobserver-1

### LINKED ISSUE #9637: Issue with Intel MPI detection
Hello,

It feels like there is a bug in Meson when trying to detect Intel MPI.

The implemented strategy is to run the MPI compiler wrapper with the `-show` option which is fine but the way the wrapper is chosen seems wrong:
https://github.com/mesonbuild/meson/blob/471c05d57dec7a6f59dcf9c80c9ed963d052320d/mesonbuild/dependencies/mpi.py#L67-L72

Meson tries to get the wrapper path or name from the `I_MPI_CC`, `I_MPI_CXX` and `I_MPI_FORT` environment values which makes no sense since according to Intel MPI documentation those should be used to "set the path/name of the underlying compiler to be used [by the wrapper]" (https://www.intel.com/content/www/us/en/develop/documentation/mpi-developer-reference-windows/top/environment-variable-reference/compilation-environment-variables.html).

That means that if `I_MPI_CC`, `I_MPI_CXX` and `I_MPI_FORT` are correctly set, Meson will try the `-show` option on the actual compiler instead of the wrapper and thus will fail to detect Intel MPI.

I would suggest using the same environment variable as the non-Intel MPI code path. I can provide a PR.

Best regards,
Rémi

### LINKED ISSUE #13615: MPI detection independent of compilers (mpich, openmpi, intel mpi) needed for conda-forge
**Describe the bug**

Meson does not detect mpich on conda-forge (using GCC) (see an example of failing build [here](https://github.com/conda-forge/fluidfft-fftwmpi-feedstock/pull/5)).

Conda-forge usually builds packages depending on MPI with different implementations. For example, in https://github.com/conda-forge/mpi4py-feedstock/blob/main/recipe/conda_build_config.yaml, there is

```
mpi:
  - mpich    # [not win]
  - openmpi  # [not win]
  - impi     # [not osx and x86_64]
  - msmpi    # [win]
```

~At least for mpich,~ the build uses GCC for mpich, openmpi and impi (see for example https://github.com/conda-forge/mpi4py-feedstock/blob/main/.ci_support/linux_64_mpiimpipython3.10.____cpython.yaml).

This issue is very similar to https://github.com/mesonbuild/meson/issues/7045 except that it is about mpich.

More generally, (I think) mpich, openmpi and Intel MPI can be compiled both with GCC and icc and this is very common to have on clusters different modules for the different cases.

**To Reproduce**

One can for example try to build https://foss.heptapod.net/fluiddyn/fluidfft/-/tree/branch/default/plugins/fluidfft-fftwmpi (simple `meson.build`) in an environment created with:

```
conda create -n env-fluidfft-mpich -c conda-forge \
  fluidfft h5py[build=mpi*] mpi4py mpich fftw[build=mpi*] cxx-compiler \
  meson meson-python pkg-config fluidfft-builder transonic cython
```

In this environment we have (as expected):

```
$ which mpicc
/data/mambaforge/envs/env-fluidfft-mpich/bin/mpicc
$ mpicc --version
x86_64-conda-linux-gnu-cc (conda-forge gcc 12.4.0-1) 12.4.0
$ mpichversion
MPICH Version:      4.2.2
MPICH Release date: Wed Jul  3 09:16:22 AM CDT 2024
MPICH ABI:          16:2:4
MPICH Device:       ch4:ofi
MPICH configure:    --prefix=/data/mambaforge/envs/env-fluidfft-mpich --disable-dependency-tracking --enable-cxx --enable-fortran --enable-f08 --with-wrapper-dl-type=none --disable-opencl --with-device=ch4
MPICH CC:           x86_64-conda-linux-gnu-cc -I/data/mambaforge/envs/env-fluidfft-mpich/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /data/mambaforge/envs/env-fluidfft-mpich/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/mpich-mpi_1722432564557/work=/usr/local/src/conda/mpich-4.2.2 -fdebug-prefix-map=/data/mambaforge/envs/env-fluidfft-mpich=/usr/local/src/conda-prefix -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /data/mambaforge/envs/env-fluidfft-mpich/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/mpich-mpi_1722432564557/work=/usr/local/src/conda/mpich-4.2.2 -fdebug-prefix-map=/data/mambaforge/envs/env-fluidfft-mpich=/usr/local/src/conda-prefix  -O2
MPICH CXX:          x86_64-conda-linux-gnu-c++ -I/data/mambaforge/envs/env-fluidfft-mpich/include -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /data/mambaforge/envs/env-fluidfft-mpich/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/mpich-mpi_1722432564557/work=/usr/local/src/conda/mpich-4.2.2 -fdebug-prefix-map=/data/mambaforge/envs/env-fluidfft-mpich=/usr/local/src/conda-prefix -O2
MPICH F77:          x86_64-conda-linux-gnu-gfortran -I/data/mambaforge/envs/env-fluidfft-mpich/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /data/mambaforge/envs/env-fluidfft-mpich/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/mpich-mpi_1722432564557/work=/usr/local/src/conda/mpich-4.2.2 -fdebug-prefix-map=/data/mambaforge/envs/env-fluidfft-mpich=/usr/local/src/conda-prefix -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /data/mambaforge/envs/env-fluidfft-mpich/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/mpich-mpi_1722432564557/work=/usr/local/src/conda/mpich-4.2.2 -fdebug-prefix-map=/data/mambaforge/envs/env-fluidfft-mpich=/usr/local/src/conda-prefix -O2
MPICH FC:           x86_64-conda-linux-gnu-gfortran -I/data/mambaforge/envs/env-fluidfft-mpich/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /data/mambaforge/envs/env-fluidfft-mpich/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/mpich-mpi_1722432564557/work=/usr/local/src/conda/mpich-4.2.2 -fdebug-prefix-map=/data/mambaforge/envs/env-fluidfft-mpich=/usr/local/src/conda-prefix -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /data/mambaforge/envs/env-fluidfft-mpich/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/mpich-mpi_1722432564557/work=/usr/local/src/conda/mpich-4.2.2 -fdebug-prefix-map=/data/mambaforge/envs/env-fluidfft-mpich=/usr/local/src/conda-prefix -O2
MPICH features:     threadcomm
```

Building fluidfft-fftwmpi with `pip install -e . -v --no-build-isolation` gives:

```
Using pip 24.2 from /data/mambaforge/envs/env-fluidfft-mpich/lib/python3.12/site-packages/pip (python 3.12)
Obtaining file:///home/users/augier3pi/dev/fluidfft/plugins/fluidfft-fftwmpi
  Running command Checking if build backend supports build_editable
  Checking if build backend supports build_editable ... done
  Running command Preparing editable metadata (pyproject.toml)
  + meson setup /home/users/augier3pi/dev/fluidfft/plugins/fluidfft-fftwmpi /home/users/augier3pi/dev/fluidfft/plugins/fluidfft-fftwmpi/build/cp312 -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=/home/users/augier3pi/dev/fluidfft/plugins/fluidfft-fftwmpi/build/cp312/meson-python-native-file.ini
  The Meson build system
  Version: 1.5.99
  Source dir: /home/users/augier3pi/dev/fluidfft/plugins/fluidfft-fftwmpi
  Build dir: /home/users/augier3pi/dev/fluidfft/plugins/fluidfft-fftwmpi/build/cp312
  Build type: native build
  Project name: fluidfft-fftwmpi
  Project version: undefined
  C++ compiler for the host machine: /data/mambaforge/envs/env-fluidfft-mpich/bin/x86_64-conda-linux-gnu-c++ (gcc 12.4.0 "x86_64-conda-linux-gnu-c++ (conda-forge gcc 12.4.0-1) 12.4.0")
  C++ linker for the host machine: /data/mambaforge/envs/env-fluidfft-mpich/bin/x86_64-conda-linux-gnu-c++ ld.bfd 2.40
  Cython compiler for the host machine: cython (cython 3.0.11)
  Host machine cpu family: x86_64
  Host machine cpu: x86_64
  Found pkg-config: YES (/data/mambaforge/envs/env-fluidfft-mpich/bin/pkg-config) 0.29.2
  Run-time dependency fftw3 found: YES 3.3.10
  mpic++ found: NO
  Run-time dependency MPI for cpp found: NO (tried config-tool)

  ../../meson.build:15:10: ERROR: Dependency "mpi" not found, tried config-tool
```

Note that using 

```
mpi_dep = dependency('mpi', language: 'cpp', method : 'config-tool')
```

does not fix the issue.

**Expected behavior**

Meson should be able to detect mpich from `mpicc` (and friends) which are in `$PATH`.

**Explanation**

The logic in https://github.com/mesonbuild/meson/blob/master/mesonbuild/dependencies/mpi.py is wrong as explained in

- https://github.com/mesonbuild/meson/issues/7045 (related old PR https://github.com/mesonbuild/meson/pull/7373)

I particularly agree with https://github.com/mesonbuild/meson/pull/7373#issuecomment-1532046963

See also if one tries to improve MPI detection:

- https://github.com/mesonbuild/meson/issues/9637 (related commit https://github.com/dguibert/meson/commit/83cd020268f03242e6c39150f7d3cddcfd163981) 

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?

Native build.

* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.)

Debian 11 or 12 (but it does should be applicable on any Unix)

* what Python version are you using e.g. 3.8.0

3.12 (but does not matter)

* what `meson --version`

1.5.99 and master

* what `ninja --version` if it's a Ninja build

1.12.1

### PR DESCRIPTION
**Pull request by paugier**
Fixes #7045, fixes #9637 and fixes #13615. See also the old PR #7373.

This PR improves support for MPI detection on Unix, which is currently quite broken. I didn't try to study what happens on Windows.

- Currently, Meson only supports IntelMPI with Intel compilers and OpenMPI with other compilers. MPICH is not supported at all. (see #7045 and #13615)

- Currently, pkg-config gets priority over detecting mpicc/mpiicc (and friends), which can lead to very unexpected results because only OpenMPI supports pkg-config and all implementations (at least on Unix) state that the correct way to detect/use MPI is to use mpicc/mpiicc (and friends). Then if the system pkg-config gives a positive result, Meson detects OpenMPI even if `PATH` has been correctly modified such that `mpicc` indicates something else.

- Environment variables `I_MPI_CC` are wrongly considered whereas they are internal to IntelMPI (see #9637).

This PR improves the situation. The detection considers (in this order):

- The standard environment variables `MPICC` (and friends)
- For Intel compilers, the commands `mpiicc` (and friends)
- The commands `mpicc` (and friends)
- Finally, `pkg-config`.

OpenMPI, MPICH, IntelMPI and other compatible implementations should be supported with different compilers. In particular OpenMPI compiled with Intel compilers and IntelMPI compiled with GCC.

MPI can be a bit tricky and there are a lot of cases to consider. It would be interesting to get the point of view of people using MPI on different clusters. CC @RemiLacroix-IDRIS, @scivision, @rcoacci, @nordmoen, @acroucher.

Note : The output of `mpicc -v` is not standardized. I tried to support what I obtain from different implementations but there might be other cases. In particular, I removed a line with `v = re.search(r'(\d{4}) Update (\d)', out)`.

### CODE REVIEW

File: mesonbuild/dependencies/mpi.py
Comment (eli-schwartz) [inline mesonbuild/dependencies/mpi.py:42]: The commit message contains extremely little content at all, and doesn't mention that this is changing the order to look for the `mpi${lang} -showme` configtools before pkg-config. Why?

File: mesonbuild/dependencies/configtool.py
Comment (eli-schwartz) [inline mesonbuild/dependencies/configtool.py:72]: CI fails because you added a function but it doesn't have type annotations.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py:42]: I was in a hurry so I quickly committed. This PR is a draft and is not ready at all. I created it only to mention that I'm working on this subject. I'm going to tell you when it is in a better state.

Comment (eli-schwartz) [reply mesonbuild/dependencies/mpi.py:42]: Ah ok, thanks for clarifying.

**Comment by RemiLacroix-IDRIS** [comment]
> Note : The output of `mpicc -v` is not standardized. I tried to support what I obtain from different implementations but there might be other cases. In particular, I removed a line with `v = re.search(r'(\d{4}) Update (\d)', out)`.

I think that's needed for older version of Intel MPI, for example here is the output I get:
```
$ mpiifort -v
mpiifort for the Intel(R) MPI Library 2019 Update 9 for Linux*
Copyright 2003-2020, Intel Corporation.
ifort version 19.1.3.304
```

**Comment by paugier** [comment]
> I think that's needed for older version of Intel MPI.

Thanks. This output should be supported now.

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py:42]: I think this is now ready to be reviewed.

File: mesonbuild/dependencies/configtool.py
Comment (paugier) [reply mesonbuild/dependencies/configtool.py]: Annotations added so that the CI is now happy.

**Comment by acroucher** [comment]
Thanks for this! I just installed your branch in a venv and tried building a minimal Fortran program that links to MPI, on an Ubuntu 22.04 machine with mpich installed. Unfortunately it didn't seem to detect MPI - not sure how much testing you've done with Fortran? (It's also possible I'm doing something in an outdated way, as I've been stuck with using Meson 0.53 since mpich detection was broken.)

Here is the program and the meson build file:

```fortran
program foo

  use mpi
  implicit none

  write(*,*) 'foo!'

end program foo
```

```meson
project('foo', 'fortran', version: '0.0.1')

mpi = dependency('mpi', language: 'fortran')
foo = executable('foo', ['foo.F90'], dependencies: [mpi])
```

and here is the relevant part of the meson-log.txt:

```
Found pkg-config: YES (/usr/bin/pkg-config) 0.29.2
Determining dependency 'ompi-fort' with pkg-config executable '/usr/bin/pkg-config'
env[PKG_CONFIG_PATH]: /home/acro018/lib/pkgconfig
env[PKG_CONFIG]: /usr/bin/pkg-config
-----------
Called: `/usr/bin/pkg-config --modversion ompi-fort` -> 1
stderr:
Package ompi-fort was not found in the pkg-config search path.
Perhaps you should add the directory containing `ompi-fort.pc'
to the PKG_CONFIG_PATH environment variable
No package 'ompi-fort' found
-----------
mpifort binary missing from cross or native file, or env var undefined.
Trying a default mpifort fallback at mpifort
Trying a default mpifort fallback at mpif90
Trying a default mpifort fallback at mpif77
mpifort found: NO
Run-time dependency MPI for fortran found: NO (tried pkgconfig, config-tool and system)

meson.build:3:6: ERROR: Dependency "mpi" not found, tried pkgconfig, config-tool and system
```

Here is the output from ``mpicc --version``:

```
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

and ``mpifort --version``:

```
GNU Fortran (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by paugier** [comment]
Thanks a lot for the testing @acroucher. I didn't try Fortran so it's not too surprising that it fails. I should be able to try this and hopefully fix this issue.

**Comment by paugier** [comment]
@acroucher can you recheck that you really use the branch [paugier:mpi-detection](https://github.com/paugier/meson/tree/mpi-detection) ? From your log, I see that pkg-config is used first, which shouldn't be the case with my branch.

I tried your small example in a conda environment created with `conda create -n env-mpich-fortran mpich fortran-compiler` and MPI is correctly detected here.

```
mpifort found: YES (/data/mambaforge/envs/env-mpich-fortran/bin/mpifort) 4.2.2
Run-time dependency MPI for fortran found: YES 4.2.2
```

Can you please also provide the output of

```
mpifort -show
# and
mpifort -v  # (only first line)
```

**Comment by acroucher** [comment]
> can you recheck that you really use the branch [paugier:mpi-detection](https://github.com/paugier/meson/tree/mpi-detection) ?

Doh! you're right, I forgot to check out the branch. I ran it again on the correct branch and it works fine. Brilliant! I also tested it on some real code and that worked too. Thank you!

File: mesonbuild/dependencies/mpi.py
Comment (eli-schwartz) [reply mesonbuild/dependencies/mpi.py:42]: I'd suggest, for reviewability, to change the order in one commit, and "support more implementations" in a second commit -- as they seem to be logically distinct changes which you are trying to make.

(I'd also like to hear more about *why* you are changing the order. :))

Comment (paugier) [reply mesonbuild/dependencies/mpi.py:42]: I split the first commit. I try to explain why mpicc has to be considered first (before pkg-config) in a commit message.

**Comment by paugier** [comment]
I did the history edition suggested. I hope I did it correctly. I get:

```
@  changeset:   16979:3dc06bdb03d3
|  bookmark:    mpi-detection
|  tag:         default/mpi-detection
|  tag:         tip
|  user:        paugier <pierre.augier@univ-grenoble-alpes.fr>
|  date:        Mon Sep 02 17:46:08 2024 +0200
|  summary:     MPI detection: get version from old IntelMPI wrappers
|
o  changeset:   16978:025928ebca77
|  parent:      16975:5d46b2165d8f
|  user:        paugier <pierre.augier@univ-grenoble-alpes.fr>
|  date:        Fri Aug 30 17:29:10 2024 +0200
|  summary:     MPI detection: support more implementations (any compilers)
|
o  changeset:   16975:5d46b2165d8f
|  user:        paugier <pierre.augier@univ-grenoble-alpes.fr>
|  date:        Wed Sep 04 23:15:56 2024 +0200
|  summary:     MPI detection: mpicc/mpiicc before pkg-config
|
o  changeset:   16974:c34a0170dda1
|  bookmark:    master
|  tag:         upstream/master
|  user:        Dylan Baker <dylan@pnwbakers.com>
|  date:        Mon Aug 19 14:29:44 2024 -0700
|  summary:     mformat: provide nice error message instead of backtrace for invalid value
```

which is quite reasonable (except maybe the dates). However, it seems that Github sorts the commits by date, which is a bit wrong here.

File: mesonbuild/dependencies/mpi.py
Comment (dcbaker) [inline mesonbuild/dependencies/mpi.py:54]: Sigh. I know what this module is doing, but if you'd like to be extra awesome, we really want to do all of these environment variable lookups in the `Environment` class, which would add the ability to look these components up from a machine file, as well as keep our guarantee that once configured Meson won't re-read and from the environment and thus change the configuration behind your back on a reconfigure.

If you'd rather not you can leave it and I'll follow up with that cleanup

**Comment by eli-schwartz** [comment]
> which is quite reasonable (except maybe the dates). However, it seems that Github sorts the commits by date, which is a bit wrong here.

That looks like mercurial output, so perhaps it's about the export format. In git on the command-line they are ordered correctly, but I also notice that the mercurial "date" field is mapped to both "AuthorDate" and "CommitDate" in git. Usually when applying/reordering commits in git, the CommitDate is the date that you performed the reordering, and I am pretty sure that's what github uses to order its history flow since it also wants to interweave commits in between the comments surrounding when the commits were made.

Weird UX issue, I guess. :) Works locally though.

File: mesonbuild/dependencies/mpi.py
Comment (eli-schwartz) [inline mesonbuild/dependencies/mpi.py]: So we no longer look up these `I_MPI_*` variables at all, just `MPI*` variables. I wonder where the original variables came from...

File: mesonbuild/dependencies/mpi.py
Comment (eli-schwartz) [inline mesonbuild/dependencies/mpi.py:189]: We now resort to -show for openmpi as well, instead of the more granular openmpi interface?

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py]: These are environment variables used internally by mpiicc (https://www.intel.com/content/www/us/en/docs/mpi-library/developer-reference-windows/2021-8/compilation-environment-variables.html).

> Set this environment variable to select a specific compiler to be used. Specify the full path to the compiler if it is not located in the search path.

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py:189]: Indeed. I chose to get a first working implementation with only `-show`, which seems to be supported by all implementations (see https://github.com/mesonbuild/meson/issues/13615#issuecomment-2320541132).

But I'm going to add a commit using `--showme` for OpenMPI and `-compile_info` / `-link_info`. It's going to be cleaner (a bit more complicated though).

**Comment by paugier** [comment]
> Finding MPI is quite a task.

Indeed! It seems to me that it is reasonable to progress by steps. Especially because it is very difficult (or impossible) to test the different possibilities.

The first step (this PR) is to fix Meson for the very common cases (in particular MPICH, OpenMPI with Intel compiler and IntelMPI with GCC).

At least, when Meson gets a good support for the common cases, more projects relying on MPI will be able to use Meson and we can expect other bug reports that will allow us to support more cases.

A first next step after this PR would be to look at what happens on Windows.

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py:189]: Done in e744f1b4523ad309e6b04c69e30ff3c1da62dd27.

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py:54]: > If you'd rather not you can leave it and I'll follow up with that cleanup

I understand your point but I don't know how to do this in Meson, so yes it would be more efficient if you follow up.

**Comment by paugier** [comment]
@eli-schwartz, is there something else I need to do on this PR? It seems to me that it really improves the situation.

**Comment by dcbaker** [comment]
Other than the loop thing, this looks super reasonable to me.

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py]: Thanks! I changed the commit to use your proposition.

**Comment by paugier** [comment]
I updated the PR. @eli-schwartz and @dcbaker should I have to do something more?

**Review by dcbaker — APPROVED** [review]

File: mesonbuild/dependencies/mpi.py
Comment (eli-schwartz) [inline mesonbuild/dependencies/mpi.py:190]: These double quotes are inconsistent and should be single quotes.

File: mesonbuild/dependencies/mpi.py
Comment (eli-schwartz) [inline mesonbuild/dependencies/mpi.py:114]: Can you document (in a comment) for future readers, which compilers accept which forms of the arguments here?

Comment (eli-schwartz) [reply mesonbuild/dependencies/mpi.py:114]: That would be e.g. mpich for -compile_info per https://github.com/mesonbuild/meson/issues/13615#issuecomment-2320541132

**Review by eli-schwartz — CHANGES_REQUESTED** [review]
> @eli-schwartz gentle ping. Can I get another feedback on this PR? Is there an issue? Should I do something more?

Thanks, this looks pretty good now. And sorry for not responding earlier, I got a bit busy and then lost track of things.

Some very minor procedural comments below, but once those are handled this looks good to merge.

**Comment by eli-schwartz** [comment]
> Fixes #7045, #9637 and #13615. See also the old PR #7373.

Aside: github unfortunately doesn't actually figure this information out on its own. It requires you to repeat the "fixes" word for each ticket if you want merging this PR to automatically close the relevant issues.

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py:115]: I added a comment.

File: mesonbuild/dependencies/mpi.py
Comment (paugier) [reply mesonbuild/dependencies/mpi.py]: I changed the double quotes as single quotes.

**Comment by paugier** [comment]
I rebased to fix a lint issue (https://github.com/mesonbuild/meson/pull/13695).

**Review by eli-schwartz — APPROVED** [review]

**Comment by eli-schwartz** [comment]
Thank you for persevering. :)

**Comment by paugier** [comment]
Thanks @eli-schwartz and @dcbaker for your reviews.

Now I like to use this on conda-forge, Spack and Guix but I will have to wait for a new Meson release.

I see that Meson 1.5.2 has just been released few days ago (https://pypi.org/project/meson/#history). Do you think we could have a new release (1.5.3 ?) in few days or weeks?

I realize that pushing a new version on PyPI for Meson might not be as straightforward as for simpler and less popular projects. So I'm just respectfully asking...

**Comment by eli-schwartz** [comment]
Unfortunately I don't think we can backport it at all. It changes the "shape" of the pickled coredata which means that existing build directories configured with meson 1.5.2 would be binary incompatible with meson 1.5.3, leading to crashes. A new major.minor release forces a full from-scratch reconfigure instead and doesn't load e.g. things like cached dependency lookups.

It's a good question though, in general, because I'm always happy to consider nominating a patch for backporting, and I'd like to get us into the situation where we issue new bugfix releases once every couple of weeks in the event that there are patches that have been nominated

**Comment by eli-schwartz** [comment]
Considering our usual release cadence, meson 1.6.0rc1 should probably be released within the month.

It's also possible to test with conda-forge and spack, if you make a VCS package available for meson. For example, in Gentoo these are called "live" ebuilds, you install "meson==9999" and it always fetches the latest code from git master. In Arch, the same thing is called "meson-git==1.5.0.r214.g1aac6cc1e"

You can also specify a dependency in pyproject.toml as:
- `meson @ git+https://github.com/mesonbuild/meson`
- `meson @ git+https://github.com/mesonbuild/meson@sha1`

if you want to guarantee that pip install and build isolation pulls in a version of meson that you know has all the features you want. pip install will anyways not preserve build directories by default, so no incremental builds of an existing worktree and therefore no worry about binary coredata compatibility.

But we really should be having a new release (candidate) within the month. The previous release was July 10, it's been 2.5 months since then, and we try to have a new release once every 3 or 4 months, which means we should have one anywhere from 2 weeks to 1.5 months from now -- and if we assume the outer limit of 1.5 months, we still need to put out release candidates at least 3 weeks ahead of time.

**Comment by eli-schwartz** [comment]
@paugier,

the plan is to tag the first release candidate next Sunday.

**Comment by eli-schwartz** [comment]
@paugier,

There is an rc1 released today and available for installation via PyPI. I packaged it in Gentoo for the benefit of people doing prerelease testing, but I don't know the policies of conda or spack around that sort of thing.

Please test this prerelease if you can, to help ensure the release occurs as smoothly as possible.

We hope to release the final 1.6.0 release in one week's time. Alas, life finds a way, and on average it's common to find at least one regression serious enough to make a second release candidate to allow people to test that fix; if this should happen, we expect the final release to happen in two week's time instead.

**Comment by paugier** [comment]
@eli-schwartz

I'm finally trying 1.6.0.rc1 and it seems that there is a problem. I'm trying to understand.

**Pull request merged**
